### PR TITLE
Fix #21484: Upkeep costs for some rides/facilities/shops are not calculated correctly

### LINKED ISSUE #21484: Abnormal & inconsistent running costs for certain rides or buildings depending on park/scenario
### Operating System

All

### OpenRCT2 build

v0.4.8 

### Base game

RollerCoaster Tycoon 2

### Area(s) with this issue?

This issue has to do with park objects and/or asset packs

### Describe the issue

Reported on Reddit: https://www.reddit.com/r/openrct2/comments/1b1r4ih/comment/ksmh0tr/

Finished the Katie's World scenario (an RCT1 scenario) a few days ago and was checking various stats for my rides and stalls before moving on to the next scenario, when I discovered this:

https://i.imgur.com/l21ppBS.jpg

Running cost for Restroom 4 is $80/hr, which is not the standard running cost for a restroom.

Here's three save files from the park.

[June, Year 1](https://drive.google.com/file/d/1AvZeqLHrE9mlm2usRB8SydfX43zIwuW_/view)
[April, Year 3](https://drive.google.com/file/d/1b_jehwnagPGs0RZcwLaD7YNOurp-y5LV/view)
[March, Year 4](https://drive.google.com/file/d/1zKRUI_VoNJ-S1ArKiW6IGVyG7CgA6zgW/view)

All three show Restroom 4 with a running cost of $80/hr.

I posted a thread about it in [r/rct](https://www.reddit.com/r/rct/), and got the following comments:

Confirmed by --[u/grumpyfan](https://www.reddit.com/user/grumpyfan/)
I downloaded your files and played with it a bit and couldn’t find anything wrong with it. I added another bathroom directly across and the running cost was $-49.60. I even deleted it and added it back, and it came back as Restroom 4 and then cost jumped again. I also opened the tile inspector but didn’t see anything different than the one I built directly across. It’s REALLY odd! Hopefully someone who knows the behind the scenes of the game more will chime in. I’m real invested now.

Also, --[u/ffsffs1](https://www.reddit.com/user/ffsffs1/)
Seems to be a bug. In my most recent Gravity Gardens save my "Toilets 4" also has a running cost of $80.00 per hour. Starting up Dusty Greens, I built 4 restrooms and while "Toilets 4" had a normal run cost, while "Toilets 3" had a run cost of $68.80 (instead of $49.60). In Factory Capers, it is the 5th restroom. I tested this 4-5 times in both maps using different locations. Meanwhile, in Crazy Castle I found no anomalies.

Moreover, in Dusty Greens if I build a ride first, the 2nd restroom will be the one that costs $68.80. If I build 2 rides first, then the first restroom will have the abnormal run cost. If I build 3 space rings, the 3rd space rings costs more than the first two. Same thing happens building 3 top spins, or 3 information kiosks. Building coasters seem to kill this pattern. For whatever reason, sometimes the nth ride/shop is guaranteed to have abnormal running costs.

All these were in openRCT2 btw.

### Steps to reproduce

1. Open Katie's World scenario. 
2. Build 4 Restrooms. 
4th restroom has an abnormal running cost of $80, compared to the others that have $49.60. 
ORCT2RR4

### Attachments

_No response_

### PR DESCRIPTION
**Pull request by ZehMatt**
There are certain fields that were not zeroed upon ride creation which however influence the upkeep cost, I believe this has been an issue for a while now and it does have an impact on a wide range of ride types.

I've checked the provided save files and the upkeep cost will change eventually to the correct thing, this might take a few seconds depending on how many rides the park has this is due to how the ride ratings are done.

Closes #21484

### CODE REVIEW

File: src/openrct2/Game.cpp [hunk 1]
```diff
@@ -481,6 +483,33 @@ static void FixInvalidSurfaces()
...
+        const auto& rtd = ride.GetRideTypeDescriptor();
+
+        // Invalid amount of drops also influence the upkeep cost.
+        if (!rtd.HasFlag(RIDE_TYPE_FLAG_HAS_DROPS) && ride.drops != 0)
+        {
+            ride.drops = 0;
+        }
+
+        // Flat rides do not have a segment length, this influences upkeep cost.
+        if (rtd.HasFlag(RIDE_TYPE_FLAG_IS_SHOP_OR_FACILITY) || rtd.HasFlag(RIDE_TYPE_FLAG_FLAT_RIDE))
```
Comment (ZehMatt) [inline src/openrct2/Game.cpp:500]: I think the assumption about flat rides here might be wrong, some ride rating tests seem to fail, or the assumption is correct and the ride rating tests always used the left over memory.

File: .clang-tidy [hunk 2]
```diff
@@ -6,6 +6,6 @@ Checks: >
     modernize-use-override
 CheckOptions:
     - key: cppcoreguidelines-macro-usage.AllowedRegexp
-      value: 'validate_global_widx'
+      value: 'validate_global_widx|NETWORK_STREAM_VERSION'
```
Comment (duncanspumpkin) [inline .clang-tidy:9]: just make network stream a constexpr

Comment (Gymnasiast) [reply .clang-tidy:9]: This is not terribly easy as the value is used in another define.

**Comment by Gymnasiast** [comment]
I’m a bit worried about the import code, especially since you are doubtful about it yourself. Unless we can verify it is absolutely correct, it might be worse than the bug it is trying to address.

As for the rest: I’m surprised we don’t zero out the ride before filling the fields. That’s going to cause a lot of subtle bugs, I fear. I’d prefer doing that in addition to explicitly setting the values like you already did.

**Comment by ZehMatt** [comment]
> I’m a bit worried about the import code, especially since you are doubtful about it yourself. Unless we can verify it is absolutely correct, it might be worse than the bug it is trying to address.
> 
> As for the rest: I’m surprised we don’t zero out the ride before filling the fields. That’s going to cause a lot of subtle bugs, I fear. I’d prefer doing that in addition to explicitly setting the values like you already did.

I agree, there is however the problem that we no longer do it the vanilla way so I don't quite know how to get a correct baseline with our adjusted ratings, I'm not even sure if this bug exists in vanilla also.

File: .clang-tidy (hunk 2 above)
Comment (ZehMatt) [reply .clang-tidy:9]: As @Gymnasiast said its used to generate the version macro so I felt like this is the better approach for the time being.

**Comment by ZehMatt** [comment]
To confirm that this fixes the issue:
1. Load the attached park.
2. Demolish Toilets 4 which is the one with upkeep of 80.
3. Create a new toilet, should be the correct upkeep cost.

[Game 1-3.zip](https://github.com/OpenRCT2/OpenRCT2/files/14469498/Game.1-3.zip)

**Review by Gymnasiast — APPROVED** [review]
I’m away today, so testing is hard, but the code looks all right, and with the import fixes postponed, I don’t think there is much potential for regression.

File: src/openrct2/actions/RideCreateAction.cpp [hunk 3]
```diff
@@ -148,6 +148,10 @@ GameActions::Result RideCreateAction::Execute() const
         station.Exit.SetNull();
         station.TrainAtStation = RideStation::NO_TRAIN;
         station.QueueTime = 0;
+        station.SegmentLength = 0;
+        station.QueueLength = 0;
+        station.Length = 0;
+        station.Height = 0;
```
Comment (janisozaur) [inline src/openrct2/actions/RideCreateAction.cpp:154]: What about
https://github.com/OpenRCT2/OpenRCT2/blob/9d77219b5d9cd3b5e2854e097f3e37ca782f3723/src/openrct2/ride/Ride.h#L55
https://github.com/OpenRCT2/OpenRCT2/blob/9d77219b5d9cd3b5e2854e097f3e37ca782f3723/src/openrct2/ride/Ride.h#L60
https://github.com/OpenRCT2/OpenRCT2/blob/9d77219b5d9cd3b5e2854e097f3e37ca782f3723/src/openrct2/ride/Ride.h#L63
?

File: src/openrct2/actions/RideCreateAction.cpp [hunk 4]
```diff
@@ -308,6 +312,7 @@ GameActions::Result RideCreateAction::Execute() const
     ride->income_per_hour = kMoney64Undefined;
     ride->profit = kMoney64Undefined;
     ride->connected_message_throttle = 0;
+    ride->drops = 0;
```
Comment (janisozaur) [inline src/openrct2/actions/RideCreateAction.cpp:315]: Can't we just `memset` the `ride` to zeros and only set non-zero values here? Seems much less error-prone

**Comment by janisozaur** [comment]
I merged this for release as I trust @ZehMatt's judgement, but my comments still hold

File: src/openrct2/actions/RideCreateAction.cpp (hunk 3 above)
Comment (ZehMatt) [reply src/openrct2/actions/RideCreateAction.cpp:154]: They are not relevant for this bug and are actually reset later.

File: src/openrct2/actions/RideCreateAction.cpp (hunk 4 above)
Comment (ZehMatt) [reply src/openrct2/actions/RideCreateAction.cpp:315]: Ride does not satisfy std::is_pod which means UB territory.

**Pull request merged**